### PR TITLE
Relax RGBA visual rules for windows

### PR DESCRIPTION
**Pull request by eviltak**
## Contents

This patch relaxes the conditions in which GTK uses the RGBA visual (i.e. enables transparency) for windows. Upstream GTK3 requires that the screen must be composited, the window manager must support `_GTK_FRAME_EXTENTS`, and the window uses CSD (client side decorations) before enabling transparency. This patch drops the latter two requirements and uses the RGBA visual for all windows so long as the screen is composited. 

For more details, see this [issue](https://gitlab.gnome.org/GNOME/gtk/-/issues/3105) and corresponding [merge request](https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/2509) in the upstream GTK3 repo; the patch introduced in this PR is equivalent to the changes introduced in the upstream merge request, which was closed due to a lack of interest from the GTK maintainers. 

I hope this is the right place for this patch; if not, feel free to close the pull request (and potentially suggest a better home for it).

## TODO

- [x] ~~I have no experience with PKGBUILDs, so this PR may require additional changes before it can be merged -- I have added the new `.patch` to the `sources` list, but some other things (such as the `sha256sums`) may require updating. The patch works correctly, but I have not tested the PKGBUILD.~~ EDIT: checksums updated (thanks [Arch Wiki](https://wiki.archlinux.org/title/Makepkg#Generate_new_checksums)!), hopefully the GitHub Action will catch any other issues.
- [ ] My limited research into the problem the patch aims to fix (mostly sourced from the linked issue) leads me to believe that the patch was not upstreamed due to CSD-related issues with `_GTK_FRAME_EXTENDS`. Perhaps we should gate the changes in this patch with a `GTK_CSD` environment variable check? So far, I have no idea what the actual problem is however (and hence no good way to test it).
- [x] I can also push a commit adding a brief description of this patch to the README, if deemed necessary.

### CODE REVIEW

**Comment by lah7** [comment]
Thanks for the pull request. The diff looks sane, but it seems risky to merge if you're not sure what it's supposed to fix, or there's no known test case of the results before and after. We should add what it helps achieve under the README's "Fixes" category if this is to go ahead.

CSD applications are already in a bad shape at the moment under GNOME's latest versions of their apps (#32)

It looks like this patch is supposed to fix that "double drop shadow" which I've seen a long time ago when Ubuntu MATE initially moved to GTK+3, but I can't remember where such bug got fixed (GTK+3, a config setting, compositor?). Is there a compositor/desktop environment out there or "classic" GTK+3 powered application that benefits from the patch?

I tried `simple-scan` under KDE without compositing, there were no shadow, as expected. Unless that's not the test environment to trigger the bug/patch.

**Comment by eviltak** [comment]
This patch essentially allows GTK-created windows to be transparent whenever a compositor is enabled. I don't think it affects shadows, because all it does is increase the color depth of the GTK-created X windows (haven't tested Wayland) to 32 bits by using the RGBA visual in a composited environment. This is required in order to have transparency; even when using a theme with transparent elements, the window itself will not appear transparent (even with a compositor running) unless it is set to the RGBA visual and `xwininfo` reports a color depth of 32 bits. 

The upstream behavior is to enable window transparency only if the three conditions listed in the PR description (compositor, `_GTK_FRAME_EXTENTS` and CSDs) are satisfied; this patch relaxes those conditions down to one: the presence of a compositor. 

The effects of this patch are apparent when you use a GTK3 theme with transparency in a setup that does not satisfy the three upstream conditions. For instance:

- `leftwm` + `picom` + upstream GTK3: created windows have a color depth of 24, i.e. no transparency, presumably because `_GTK_FRAME_EXTENTS` is not supported. 
-  `leftwm` + `picom` + patched GTK3: created windows have a color depth of 32 allowing them to be transparent, as the only required condition of a compositor being running is satisfied. 
- Plasma + KWin compositor + Upstream GTK3: created windows support transparency out of the box 
- Plasma + `picom` + Upstream GTK3: no window transparency, requires this patch for GTK windows to have transparency. 

This patch should not affect setups without a running compositor.

TL;DR: Upstream GTK3 disables window transparency in some cases even though it is supported (only requirement should be a compositor); this patch enables transparency in such situations. I hope that makes the problem this patch attempts to solve more clear -- apologies for not including enough detail in the original PR description.

**Review by lah7 — APPROVED** [review]
Thanks for the explanation. I can see a _crude_ example of it working by adding this CSS to `~/.config/gtk-3.0/gtk.css`:
```css
* {
    background-color: rgba(0,0,0,0.2);
}
```
Sure enough, many elements of the window became translucent black when the compositor is running, that otherwise would've been opaque under regular GTK3. Let's merge this!

**Comment by eviltak** [comment]
Thanks a lot! Thanks for committing the README addition as well.
